File: src/main/java/io/github/huacnlee/autocorrectIdeaPlugin/Formatter.java
package io.github.huacnlee.autocorrectIdeaPlugin;

import com.intellij.openapi.command.WriteCommandAction;
import com.intellij.openapi.editor.Document;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.editor.SelectionModel;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.Project;
import io.github.huacnlee.AutoCorrect;

public class Formatter {
    // https://plugins.jetbrains.com/docs/intellij/working-with-text.html#safely-replacing-selected-text-in-the-document
    public static void format(Project project, Document doc) {
        final String filepath = FileDocumentManager.getInstance().getFile(doc).getPath();
        Config config = Config.getCurrent(project.getBasePath());
        if (config.isIgnored(filepath)) {
            System.out.printf("%s is ignored.\n", filepath);
            return;
        }

        String text = doc.getText();

        final String newText = AutoCorrect.formatFor(text, filepath);
        WriteCommandAction.runWriteCommandAction(project, () -> {
            int startOffset = 0;
            int endOffset = doc.getTextLength();
            doc.replaceString(startOffset, endOffset, newText);
        });
    }

}

File: src/main/java/io/github/huacnlee/autocorrectIdeaPlugin/Formatter.kt
package io.github.huacnlee.autocorrectIdeaPlugin

import com.intellij.openapi.command.WriteCommandAction
import com.intellij.openapi.editor.Document
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.project.Project
import io.github.huacnlee.AutoCorrect

object Formatter {
    // https://plugins.jetbrains.com/docs/intellij/working-with-text.html#safely-replacing-selected-text-in-the-document
    @JvmStatic
    fun format(project: Project, doc: Document) {
        var file = FileDocumentManager.getInstance().getFile(doc);
        val filepath = FileDocumentManager.getInstance().getFile(doc)!!.path
        val config = Config.getCurrent(project.basePath)
        if (config.isIgnored(filepath)) {
            System.out.printf("%s is ignored.\n", file!!.name)
            return
        }
        val text = doc.text
        val newText = AutoCorrect.formatFor(text, filepath)
        WriteCommandAction.runWriteCommandAction(project) {
            val startOffset = 0
            val endOffset = doc.textLength
            doc.replaceString(startOffset, endOffset, newText)
        }
    }
}